tests/unit/test_topology.py

```diff
@@ -0,0 +1,112 @@
+from copy import deepcopy
+
+import pytest
+
+from expertsystem.topology import (
+    Edge,
+    InteractionNode,
+    SimpleStateTransitionTopologyBuilder,
+    Topology,
+)
+
+
+class TestEdge:
+    @staticmethod
+    def test_get_connected_nodes():
+        edge = Edge(1, 2)
+        assert edge.get_connected_nodes() == {1, 2}
+        edge = Edge(originating_node_id=3)
+        assert edge.get_connected_nodes() == {3}
+        edge = Edge(ending_node_id=4)
+        assert edge.get_connected_nodes() == {4}
+
+
+class TestTopology:
+    @staticmethod
+    def test_constructor_exceptions():
+        with pytest.raises(ValueError):
+            assert Topology(edges={0: Edge(None, 1)})
+        with pytest.raises(ValueError):
+            assert Topology(nodes={0}, edges={0: Edge(1, None)})
+        with pytest.raises(ValueError):
+            assert Topology(edges={0: Edge(None, None)})
+        topology = Topology(nodes={0}, edges={0: Edge(0, None)})
+        assert len(topology.nodes) == 1
+
+    @staticmethod
+    def test_is_valid(dummy_topology):
```

Comment:
I would rather make this into a test that creates various topologies as inputs as this dummy topology, for example using the TopologyBuilder. Because its very hard to understand this test if you dont know what the dummy topology is.

Reply:
41cc7c1b